packages/yarnpkg-pnp/sources/esm-loader/loader.ts

```diff
@@ -0,0 +1,192 @@
+import fs                                  from 'fs';
+import moduleExports                       from 'module';
+import path                                from 'path';
+import {fileURLToPath, pathToFileURL, URL} from 'url';
+
+import * as nodeUtils                      from '../loader/nodeUtils';
+import {PnpApi}                            from '../types';
+
+function isValidURL(str: string) {
+  try {
+    new URL(str);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+const builtins = new Set([...moduleExports.builtinModules]);
+
+const pathRegExp = /^(?![a-zA-Z]:[\\/]|\\\\|\.{0,2}(?:\/|$))((?:node:)?(?:@[^/]+\/)?[^/]+)\/*(.*|)$/;
+
+async function exists(path: string) {
+  try {
+    await fs.promises.access(path, fs.constants.R_OK);
+    return true;
+  } catch { }
+  return false;
+}
+
+export async function resolve(
+  originalSpecifier: string,
+  context: any,
+  defaultResolver: any
+) {
+  const {findPnpApi} = (moduleExports as unknown as { findPnpApi?: (path: NativePath) => null | PnpApi });
+  if (!findPnpApi)
+    return defaultResolver(originalSpecifier, context, defaultResolver);
+
+  let specifier = originalSpecifier;
+  let validURL: boolean | undefined;
+  if (builtins.has(specifier) || (validURL = isValidURL(specifier))) {
+    if (!validURL || pathToFileURL(specifier).protocol !== `file:`) {
+      return defaultResolver(originalSpecifier, context, defaultResolver);
+    } else {
+      specifier = fileURLToPath(specifier);
+    }
+  }
+
+  const {parentURL, conditions = []} = context;
+
+  const issuer = parentURL ? fileURLToPath(parentURL) : process.cwd();
+
+  const pnpapi = findPnpApi(issuer) ?? (validURL ? findPnpApi(specifier) : null);
+  if (!pnpapi)
+    return defaultResolver(originalSpecifier, context, defaultResolver);
+
+  const dependencyNameMatch = specifier.match(pathRegExp);
+
+  let allowLegacyResolve = false;
+
+  if (dependencyNameMatch) {
+    const [, dependencyName, subPath] = dependencyNameMatch as [unknown, string, PortablePath];
+
+    // https://github.com/nodejs/node/blob/0996eb71edbd47d9f9ec6153331255993fd6f0d1/lib/internal/modules/esm/resolve.js#L686-L691
+    if (subPath === ``) {
+      const resolved = pnpapi.resolveToUnqualified(`${dependencyName}/package.json`, issuer);
+      if (resolved && await exists(resolved)) {
+        const pkg = JSON.parse(await fs.promises.readFile(resolved, `utf8`));
+        allowLegacyResolve = pkg.exports == null;
+      }
+    }
+  }
+
+  const result = pnpapi.resolveRequest(specifier, issuer, {
+    conditions: new Set(conditions),
+    // TODO: Handle --experimental-specifier-resolution=node
+    extensions: allowLegacyResolve ? undefined : [],
+  });
+
+  if (!result)
+    throw new Error(`Resolution failed`);
+
+  return {
+    url: pathToFileURL(result).href,
+  };
+}
+
+const realModules = new Set<string>();
+
+export async function getFormat(
+  resolved: string,
+  context: any,
+  defaultGetFormat: any
+) {
```

Comment:
Why do we have to reimplement this? (Worth a comment in the code as well)

Reply:
See https://github.com/yarnpkg/berry/pull/2161#discussion_r548892218 (Will add a comment though)